### PR TITLE
Refactor settings access to use WC_Stripe_Settings singleton

### PR DESCRIPTION
**Pull request by wjrosa**
## Changes proposed in this Pull Request:

Refactors settings access throughout the plugin to use a `WC_Stripe_Settings` singleton class with typed getters and setters, replacing direct array access via `WC_Stripe_Helper::get_stripe_settings()`.

### What this PR does

- **Introduces `WC_Stripe_Settings`** (`includes/class-wc-stripe-settings.php`) — a singleton that wraps the `woocommerce_stripe_settings` option with:
  - **40+ typed getters** (e.g. `is_enabled()`, `is_testmode()`, `get_publishable_key()`, `is_capture_enabled()`, `are_keys_set()`)
  - **Setters** for commonly written keys (e.g. `set_pmc_enabled()`, `set_apple_pay_domain_set()`)
  - **Generic accessors** (`get($key, $default)`, `set($key, $value)`, `has($key)`, `delete($key)`, `save()`) for bulk mutation paths
  - **Auto-invalidation** via `update_option` hook — singleton resets when the DB option is updated externally
  - **Public `reset()` method** for test isolation

- **Migrates 30+ files** from `WC_Stripe_Helper::get_stripe_settings()` raw array access to typed `WC_Stripe_Settings` methods across:
  - Core services (`WC_Stripe_API`, `WC_Stripe_Mode`, `WC_Stripe_Logger`, `WC_Stripe_Webhook_Handler/State`)
  - Payment gateways (`UPE_Payment_Gateway`, `UPE_Payment_Method`, voucher gateway)
  - Express checkout (`Express_Checkout_Element`, `Express_Checkout_Helper`)
  - Admin controllers (settings, account keys, OC toggle, payment gateways, payment requests, express checkout, Amazon Pay, inbox notes, admin notices)
  - Connect / OAuth (`WC_Stripe_Connect`)
  - Feature flags (`WC_Stripe_Feature_Flags`)
  - Apple Pay registration, blocks support, agentic commerce
  - Helper class (`WC_Stripe_Helper`)
  - Promotion notes (BnPL, OC)

- **Fixes bugs**:
  - `WC_Stripe_Mode` was reading wrong `test_mode` key instead of `testmode` — now uses `is_testmode()`
  - `statement_descriptor` property incorrectly typed as `bool` → `string` (in both UPE gateway and voucher abstract)
  - Webhook handler `$secret` property typed as `string` → `string|false`

- **Adds `WC_Stripe_Helper::is_connected()`** convenience method

### Benefits

- Type-safe settings access with IDE autocompletion and PHPStan validation
- Centralized default values — no more scattered `isset()` / `?? ''` checks
- Single source of truth for settings reads and writes
- Automatic cache invalidation when settings change (via WordPress option hooks)
- Easier to test — `WC_Stripe_Settings::reset()` provides clean test isolation

### Automated
1. Run `npm run phpstan` — should pass with no errors
2. Run `npm run test:php` — `WC_Stripe_Settings_Test` (13 test methods) should pass. Note: a small number of pre-existing test failures from the develop merge (PMC `is_enabled` change, agentic commerce order mapper) are unrelated to this PR

### Settings page
1. Go to **WooCommerce > Settings > Payments > Stripe**
2. Verify the settings page loads without errors
3. Toggle the **Enable Stripe** checkbox, save, and verify it persists
4. Toggle **Test mode**, save, and verify it persists
5. Enter test API keys (publishable and secret), save, verify they persist
6. Enable/disable **Express Checkout**, verify the setting saves correctly
7. Change express checkout button locations, verify they save

### Payment processing
1. With test keys configured, place a test order using a card payment
2. Verify the payment completes successfully
3. If express checkout is enabled, verify Apple Pay / Google Pay buttons appear on product and cart pages

### OAuth connect flow
1. If possible, disconnect and reconnect the Stripe account via OAuth
2. Verify keys, connection type, and PMC settings are saved correctly after reconnect

### Webhook configuration
1. Verify webhook status displays correctly on the settings status page
2. If webhooks are configured, verify they continue to process events

### Edge cases
1. Verify the plugin activates cleanly on a fresh install (no pre-existing settings)
2. Verify the Express Checkout disabled/enabled state shows correctly on the system status page

---

- [x] Covered with tests (`WC_Stripe_Settings_Test` with 13 test methods covering all getters, setters, defaults, edge cases)
- [ ] Tested on mobile (does not apply — backend refactor only)

### Changelog entry

* Dev - Refactor settings access to use WC_Stripe_Settings singleton with typed getters and setters instead of raw array access

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by wjrosa** [comment]
@woocommerce/quark I tried to update the settings retrieval code with the main `WC_Stripe` class (as @diegocurbelo suggested during the meetup). See some of the results [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4408/files). That did not work because some classes we instantiate inside `WC_Stripe` (like `WC_Stripe_Payment_Request`) require the settings. That was causing an infinite loop. The solution would be to access the options on those classes directly, but that would leave us with two different standards anyway.

Today, I came up with this solution here instead. Although it introduces a new class, it uses the singleton pattern and also guarantees that settings are always fresh when updated. What do you think?

This is just a POC PR. If you think the approach is nice, we can work on this later with multiple smaller PRs.

File: includes/class-wc-stripe-apple-pay-registration.php [hunk 1]
```diff
@@ -195,10 +195,10 @@ public function register_domain( $secret_key ) {
...
 
 		} catch ( Exception $e ) {
-			$settings                              = WC_Stripe_Helper::get_stripe_settings();
-			$settings['apple_pay_verified_domain'] = $this->domain_name;
-			$settings['apple_pay_domain_set']      = 'no';
-			WC_Stripe_Helper::update_main_stripe_settings( $settings );
+			$settings    = WC_Stripe_Settings::get_instance();
+			$domain_name = (string) $this->domain_name;
+			$settings->set_apple_pay_verified_domain( $domain_name );
+			$settings->set_apple_pay_domain_set( 'no' );
```
[the remark this answers is not in the document]
Comment (wjrosa) [reply includes/class-wc-stripe-apple-pay-registration.php]: Fixed in edb02b42 — removed the `set_apple_pay_verified_domain` call from the catch path so transient failures no longer prevent retries.

File: includes/class-wc-stripe-settings.php [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$settings = get_option( self::SETTINGS_OPTION, [] );
+		if ( ! is_array( $settings ) ) {
+			$settings = [];
+		}
+		$this->settings = $settings;
+	}
```
[the remark this answers is not in the document]
Comment (wjrosa) [reply includes/class-wc-stripe-settings.php:57]: Fixed in edb02b42 — constructor is now `private`.

File: includes/payment-methods/class-wc-stripe-express-checkout-helper.php [hunk 3]
```diff
@@ -17,7 +17,7 @@ class WC_Stripe_Express_Checkout_Helper {
 	/**
 	 * Stripe settings.
 	 *
-	 * @var array
+	 * @var WC_Stripe_Settings
 	 */
 	public $stripe_settings;
```
[the remark this answers is not in the document]
Comment (wjrosa) [reply includes/payment-methods/class-wc-stripe-express-checkout-helper.php:22]: Fixed in f3e2d268 — the tests that stubbed `stripe_settings` as an array now use `WC_Stripe_Settings::get_instance()` after setting up settings via the DB. The property type is `WC_Stripe_Settings` with auto-invalidation on option updates, so external callers that update settings via `WC_Stripe_Helper::update_main_stripe_settings()` will get fresh reads.

File: includes/class-wc-stripe-settings.php [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+		if ( ! is_array( $settings ) ) {
+			$settings = [];
+		}
+		$this->settings = $settings;
+
+		// Invalidate the singleton when the option is updated externally (e.g. by WP update_option).
+		if ( ! self::$hook_registered ) {
+			add_action( 'update_option_' . self::SETTINGS_OPTION, [ __CLASS__, 'reset' ] );
+			self::$hook_registered = true;
+		}
```
[the remark this answers is not in the document]
Comment (wjrosa) [reply includes/class-wc-stripe-settings.php]: Fixed in edb02b42 — added `add_action( 'delete_option_' . self::SETTINGS_OPTION, [ __CLASS__, 'reset' ] )` alongside the existing `update_option_` hook.

File: includes/class-wc-stripe-settings.php [hunk 5]
```diff
@@ -0,0 +1,768 @@
...
+	/**
+	 * Get the value of the `upe_checkout_experience_accepted_payments` setting.
+	 *
+	 * @return array<int, string>
+	 */
+	public function get_upe_checkout_experience_accepted_payments(): array {
+		return ! empty( $this->settings['upe_checkout_experience_accepted_payments'] )
+			? $this->settings['upe_checkout_experience_accepted_payments']
+			: [ WC_Stripe_Payment_Methods::CARD ];
+	}
```
[the remark this answers is not in the document]
Comment (wjrosa) [reply includes/class-wc-stripe-settings.php:338]: Fixed in edb02b42 — now validates with `is_array()` before returning, falling back to `[WC_Stripe_Payment_Methods::CARD]` for any non-array or empty value.

File: includes/connect/class-wc-stripe-connect.php [hunk 6]
```diff
@@ -263,37 +263,53 @@ private function save_stripe_keys( $result, $type = 'connect', $mode = 'live' )
...
+			$is_test         = 'live' !== $mode;
+			$prefix          = $is_test ? 'test_' : '';
+			$default_options = $this->get_default_stripe_config();
+			// Ensure defaults are applied to the settings singleton.
+			$settings = WC_Stripe_Settings::get_instance();
+			foreach ( $default_options as $key => $value ) {
+				if ( ! $settings->has( $key ) ) {
+					$settings->set( $key, $value );
+				}
+			}
```
[the remark this answers is not in the document]
Comment (wjrosa) [reply includes/connect/class-wc-stripe-connect.php]: Fixed in edb02b42 — now captures `$previous_pmc_enabled = $settings->get_pmc_enabled()` before applying defaults and setting `pmc_enabled`, and uses the captured value for the migration guard.

File: tests/phpunit/class-wc-stripe-settings-test.php [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+	 * Test constructor handles non-array option gracefully.
+	 */
+	public function test_constructor_handles_non_array_option() {
+		// Remove the filter that expects an array value before setting a non-array option.
+		remove_filter( 'pre_update_option_' . WC_Stripe_Settings::SETTINGS_OPTION, [ WC_Stripe::get_instance(), 'gateway_settings_update' ] );
+		update_option( WC_Stripe_Settings::SETTINGS_OPTION, 'invalid' );
+		WC_Stripe_Settings::reset();
+		$settings = WC_Stripe_Settings::get_instance();
+		$this->assertFalse( $settings->is_enabled() );
+	}
```
[the remark this answers is not in the document]
Comment (wjrosa) [reply tests/phpunit/class-wc-stripe-settings-test.php:170]: Fixed in edb02b42 — wrapped in try/finally that restores the filter via `add_filter()`.

**Comment by wjrosa** [comment]
Closing in favor of a 3-part stacked PR series for easier review: see linked PRs below.

**Comment by wjrosa** [comment]
Split into 3 stacked PRs for easier review:

1. **#5265** — Introduce `WC_Stripe_Settings` class (new file + tests, no migration)
2. **#5266** — Migrate core services + payment methods
3. **#5267** — Migrate admin + Connect + remaining consumers
